ext/semian/simple_integer.c

```diff
@@ -0,0 +1,90 @@
+#include "simple_integer.h"
+
+#include <errno.h>
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include "types.h"
+#include "util.h"
+
+static const rb_data_type_t semian_simple_integer_type;
```

Comment:
We don't have a `free` function, so would we be leaking `shmget` memory that is no longer in use?

Similarly for `circuit_breaker.c.` and `sliding_window.c`.

Reply:
Yes, this is one of the outstanding issues with this PR, along with the race conditions.
@spike01 is working on the race conditions, but I don't have a solution for the memory leak right now.

Reply:
You can set `semian_simple_integer_type.function.dfree` to your free function:
```c
  typedef struct rb_data_type_struct rb_data_type_t;

  struct rb_data_type_struct {
          const char *wrap_struct_name;
          struct {
                  void (*dmark)(void*);
                  void (*dfree)(void*);
                  size_t (*dsize)(const void *);
                  void *reserved[2];
          } function;
          const rb_data_type_t *parent;
          void *data;
          VALUE flags;
  };
```

```
wrap_struct_name is an identifier of this instance of the struct.
It is basically used for collecting and emitting statistics.
So the identifier must be unique in the process, but doesn't need
to be valid as a C or Ruby identifier.

These dmark / dfree functions are invoked during GC execution.  No
object allocations are allowed during it, so do not allocate ruby
objects inside them.

dmark is a function to mark Ruby objects referred from your struct.
It must mark all references from your struct with rb_gc_mark or
its family if your struct keeps such references.

--
Note that it is recommended to avoid such a reference.
++

dfree is a function to free the pointer allocation.
If this is -1, the pointer will be just freed.
```

Reply:
Thanks for the tip!
I updated the code to use the `dfree` functions to clean up.